Only create epochs if there is something to process or the maximum nu…

Liamsi opened the pull request:
…mber of min epoch intervals has passed

- rename flag period to min-period
- add max-period flag (number of passed min-periods until signer creates epoch independent from mutations)
- add flag to `CreateEpoch` to let the caller enforce creating a new epoch (independent from mutations); update tests
resolves #509

Liamsi commented:
I signed it!

cesarghali commented on cmd/keytransparency-signer/backend.go:
```diff
@@ -45,7 +45,8 @@ import (
 	domain        = flag.String("domain", "example.com", "Distinguished name for this key server")
-	epochDuration = flag.Duration("period", time.Second*60, "Time between epoch creation")
+	minEpochDuration = flag.Duration("min-period", time.Second*60, "Minimum time between epoch creation (create epochs only if there where mutations)")
+	maxElapsedEpochs = flag.Uint("max-period", 60, "Maximum elapsed min-periods until epoch creation (independent from mutations)")
```
Can this also be `flag.Duration` like `minEpochDuration`?

cesarghali replied:
Can this also be `flag.Duration` like `minEpochDuration`?

Liamsi replied:
Done (in https://github.com/google/keytransparency/pull/593/commits/08bac3f7af45be342be3a6b758f949e5d4aebefe)

gdbelvin commented on core/signer/signer.go:
```diff
@@ -145,7 +157,7 @@ func (s *Signer) applyMutations(mutations []*tpb.SignedKV, leaves []*trillian.Ma
 
 // CreateEpoch signs the current map head.
-func (s *Signer) CreateEpoch(ctx context.Context) error {
+func (s *Signer) CreateEpoch(ctx context.Context, alwaysCreateNewEpoch bool) error {
```
go style prefers shorter names.  Consider `forceNewEpoch`?

Liamsi replied:
Done (in https://github.com/google/keytransparency/pull/593/commits/5987035a462e7f080285f51e879b117ae5e2b2b1)

cesarghali commented on core/signer/signer.go:
```diff
@@ -64,9 +64,21 @@ func New(realm string,
+		var createNewEpoch bool
+		if elapsed >= maxElapsed {
+			createNewEpoch = true
+			elapsed = 0
```
How about we use a time here for this instead of a counter, this way the unit of `maxElapsed` is not coupled with `minInterval`? What do you think?

Liamsi replied:
Thanks for the suggestion! I coupled both like above because I wanted to avoid creating two independent "threads"/go-routines that might "race" for the epoch creation. But your suggestions is simpler and avoid this, too (if I understood it correctly).

cesarghali commented:
Also please run `git rebase master` to update the branch. Thanks.

Liamsi commented:
Thanks a lot for your reviews @gdbelvin @cesarghali. I made some changes and I think this might be ready for another review. 

> Please add a test

I've added one that tests the epoch creation logic (counts if calling `CreateEpoch` with `forceNewEpoch` as expected). 
Another thing that might be worthwhile to test is if epochs in between `maxDurations` are really only created if there are mutations. I didn't include that in this PR as one needs to fire up quite a lot to have a `Signer` running. Maybe this is sth. which could into some `integration` package.

gdbelvin commented on cmd/keytransparency-signer/backend.go:
```diff
@@ -132,7 +133,11 @@ func main() {
 	mutator := entry.New()
 
+	if *maxEpochDuration < *minEpochDuration {
+		glog.Exitf("Maximum time between epoch creation (%v) must be grater or equal minimum time between epoch creation.")
```
Consider following the canonical "got, want" style for error messages. 
https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures

gdbelvin replied:
It's only required for tests, but it's useful info to have for other failures too because it also gives the desired value.

cesarghali replied:
SG.

gdbelvin commented on core/signer/signer.go:
```diff
@@ -63,10 +63,26 @@ func New(realm string,
+	processEpochs(ctx, tc, maxInterval, s.CreateEpoch)
+}
+
+func processEpochs(ctx context.Context, tc <-chan time.Time, maxElapsed time.Duration, sign func(ctx context.Context, enforce bool) error) {
```
please add a function comment

cesarghali replied:
Actually, do we need this function? `StartSigning` just calls `processEpochs`.

Liamsi replied:
The only reason I only created `processEpochs` to make it possible to test the underlying logic without the need of creating a whole `Signer` (which depends on a lot of other running objects, like the DB, a mutator, transaction factory...). Other than making testing convinient we don't need, I guess.

gdbelvin commented on core/signer/signer.go:
```diff
@@ -63,10 +63,26 @@ func New(realm string,
+}
+
+func processEpochs(ctx context.Context, tc <-chan time.Time, maxElapsed time.Duration, sign func(ctx context.Context, enforce bool) error) {
+	last := time.Now()
```
fetch time last from previous map head if it exists?

gdbelvin commented on core/signer/signer.go:
```diff
@@ -63,10 +63,26 @@ func New(realm string,
+		} else {
+			forceNewEpoch = false
+		}
+		// only create a new epoch if elapsed intervals >= max. elapsed intervals
```
s/only/Only/

gdbelvin commented on core/signer/signer_test.go:
```diff
@@ -0,0 +1,53 @@
+	"golang.org/x/net/context"
+)
+
+func TestEpochCreation(t *testing.T) {
```
Prefer table driven tests if at all possible. 
https://github.com/golang/go/wiki/TableDrivenTests

gdbelvin commented on core/signer/signer_test.go:
```diff
@@ -0,0 +1,53 @@
+	if got != want {
+		t.Fatalf("Expected %v maxEpoch(s) but got %v", want, got)
+	}
+}
```
add newline

cesarghali commented on core/signer/signer.go:
```diff
@@ -63,10 +63,26 @@ func New(realm string,
+			forceNewEpoch = false
+		}
+		// only create a new epoch if elapsed intervals >= max. elapsed intervals
+		// OR there were mutations in between max. elapsed epochs (expected to be handled by passed callback):
```
Although there's no line limit in Go, we're trying to keep comments to 80 characters a line for better readability.

cesarghali commented on core/signer/signer.go:
```diff
@@ -63,10 +63,26 @@ func New(realm string,
+		}
+		// only create a new epoch if elapsed intervals >= max. elapsed intervals
+		// OR there were mutations in between max. elapsed epochs (expected to be handled by passed callback):
+		if err := sign(ctx, forceNewEpoch); err != nil {
```
Should this be a call to `CreateEpoch`? I can't find `sign` function.

Liamsi replied:
Indeed, this is a call to `CreateEpoch`. `sign` is a "callback" passed to `processEpochs` (and `s.CreateEpoch` is passed as an argument).
Again, this only exists to make it testable without having to run a whole environment (like for instance in this [integration](https://github.com/google/keytransparency/blob/master/integration/testutil.go#L154-L213) package).

cesarghali replied:
That makes sense, thanks for the explanation.

gdbelvin commented on core/signer/signer.go:
```diff
@@ -44,13 +44,13 @@ type Signer struct {
-	mutator mutator.Mutator,
-	mutations mutator.Mutation,
-	factory transaction.Factory) *Signer {
+mapID int64,
```
Make sure to run gofmt. Looks like there's a tab missing.

gdbelvin replied:
Make sure to run gofmt. Looks like there's a tab missing.

cesarghali replied:
Running `make presubmit` would take care of all of that and shows other possible errors :)

gdbelvin commented on core/signer/signer.go:
```diff
@@ -63,10 +63,66 @@ func New(realm string,
-		if err := s.CreateEpoch(ctx); err != nil {
+// StartSigning advance epochs once per minInterval, if there were mutations,
+// and at least once per maxElapsed minIntervals.
+func (s *Signer) StartSigning(ctx context.Context, minInterval time.Duration, maxInterval time.Duration) {
```
You can shorten the function signature by dropping repeated types
`ctx context.Context, minInterval, maxInterval time.Duration`

gdbelvin commented on core/signer/signer.go:
```diff
@@ -63,10 +63,66 @@ func New(realm string,
+// StartSigning advance epochs once per minInterval, if there were mutations,
+// and at least once per maxElapsed minIntervals.
+func (s *Signer) StartSigning(ctx context.Context, minInterval time.Duration, maxInterval time.Duration) {
+	// TODO Fetch last time from previous map head (as stored in the map server) if it exists:
```
TODO's should have the form: 
TODO(username):

gdbelvin commented on core/signer/signer.go:
```diff
@@ -63,10 +63,66 @@ func New(realm string,
+}
+
+// processEpochs calls the given callback (usually Signer.CreateEpoch) every
+// minElapsed and every maxElapsed time.
```
Collapse comments into solid blocks 80 chars wide. 
in vim: `gq`

gdbelvin commented on core/signer/signer.go:
```diff
@@ -63,16 +63,65 @@ func New(realm string,
+	}
+	// Start issuing epochs:
+	enforceChan := processEpochs(last, minInterval, maxInterval)
+	for f := range enforceChan {
```
`for f := range epochTicker(last, minInterval, maxInterval)` ?

gdbelvin commented on core/signer/signer.go:
```diff
@@ -63,16 +63,65 @@ func New(realm string,
+// processEpochs returns and sends to a bool channel every time an epoch should
+// be created. If the boolean value is true this indicates that the epoch should
+// be created independent from if there were mutations or not.
+func processEpochs(last time.Time, minElapsed, maxElapsed time.Duration) <-chan bool {
```
processesEpochs is a slightly confusing function name since it no longer processes epochs.  :-)
Consider epochTicker or another term that indicates that it outputs Ticks

gdbelvin commented on core/signer/signer.go:
```diff
@@ -63,16 +63,65 @@ func New(realm string,
+		}
+
+		for now := range tc {
+			if (time.Since(last) + minElapsed) >= maxElapsed {
```
We should be consistent for what value of "now" to use. 
Consider using the `now` variable from the for loop rather than the embedded `time.Now()` inside `Since`

Liamsi replied:
Good catch! Changed.

Liamsi commented:
Thanks for the review! I addressed your comments in https://github.com/google/keytransparency/pull/593/commits/2b9b2b432ba646f03beff446c78f86f264675a65.
